jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java

```diff
@@ -0,0 +1,170 @@
+ * <li>Twice the cost of a perfect matching in a graph is less than or equal to the cost of every Hamiltonian
+ * cycle. This follows from the fact that after forming a closed tour using the edges of a perfect matching
+ * the cost of the edges not from the matching is greater than or equal to the cost of the matching edges.</li>
+ * </ul>
+ * <p>
+ * For more details, see <i>Christofides, Nicos. (1976). Worst-Case Analysis of a New Heuristic for the Traveling
+ * Salesman Problem. Carnegie Mellon University. 10.</i>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Timofey Chudakov
+ * @author Dimitrios Michail
+ * @since September 2018
+ */
+public class ChristofidesThreeHalvesApproxMetricTSP<V, E> implements HamiltonianCycleAlgorithm<V, E> {
+
+    /**
+     * Empty constructor
+     */
+    public ChristofidesThreeHalvesApproxMetricTSP() {
+    }
+
+    /**
+     * Computes a $3/2$-approximate tour.
+     *
+     * @param graph the input graph
+     * @return a tour
+     * @throws IllegalArgumentException if the graph is not undirected
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        if (!graph.getType().isUndirected()) {
+            throw new IllegalArgumentException("Graph must be undirected");
+        }
+        if (!GraphTests.isComplete(graph)) {
+            throw new IllegalArgumentException("Graph is not complete");
+        }
+        if (graph.vertexSet().isEmpty()) {
+            throw new IllegalArgumentException("Graph contains no vertices");
+        }
+
+        /*
+         * Special case singleton vertex
+         */
+        if (graph.vertexSet().size() == 1) {
+            V start = graph.vertexSet().iterator().next();
+            return new GraphWalk<>(
+                    graph, start, start, Collections.singletonList(start), Collections.emptyList(), 0d);
+        }
+
+        int n = graph.vertexSet().size();
+        // add all vertices of the graph to the auxiliary graph
+        Graph<V, DefaultEdge> mstAndMatching = new Pseudograph<>(DefaultEdge.class);
+        graph.vertexSet().forEach(mstAndMatching::addVertex);
+
+        // add all edges of a minimum spanning tree to the auxiliary graph
+        SpanningTreeAlgorithm<E> spanningTreeAlgorithm = new KruskalMinimumSpanningTree<>(graph);
+        spanningTreeAlgorithm.getSpanningTree().getEdges().forEach(
+                e -> mstAndMatching.addEdge(graph.getEdgeSource(e), graph.getEdgeTarget(e)));
+
+        // find odd degree vertices
+        Set<V> oddDegreeVertices = new HashSet<>();
+        mstAndMatching.vertexSet().forEach(v -> {
+            if ((mstAndMatching.edgesOf(v).size() & 1) == 1) { // if degree is odd
+                oddDegreeVertices.add(v);
+            }
+        });
```

Comment:
You can write this as `Set<V> oddDegreeVertices = mstAndMatching.vertexSet().stream().filter(v -> (mstAndMatching.edgesOf(v).size() & 1) == 1).collect(Collectors.toSet());`.